### ISSUE TITLE
Icons too small and other scaling problems on high DPI monitor

### ISSUE THREAD

**Issue by arencambre** [qgis/QGIS#40274]
**Describe the bug**

I have an HP EliteBook with a 4K (high DPI) monitor. When I open QGIS, the icons are tiny (Chrome browser window included in screenshot for context, and it may appear to be much larger than reality due to how Github apparently uses 72 dpi for displaying images):
![image](https://user-images.githubusercontent.com/411495/100234466-c1008e80-2ef0-11eb-8ec1-1796b57e25ae.png)

This is a fresh install, through osgeo4w.

The **Help** --> **About** menu comes up badly messed up:
![image](https://user-images.githubusercontent.com/411495/100234679-0a50de00-2ef1-11eb-9e91-7f99315dd767.png)

**How to Reproduce**

1. On a computer with a 4K monitor, open QGIS. Notice how small the icons are.
2. **Help** --> **About**

**Expected behavior**

QGIS and its features scale appropriately for high DPI monitors without any intervention on the part of users. I am sure some will post workarounds, and those will help others finding this bug, but workarounds are not bug fixes.

**QGIS and OS versions**

QGIS version
3.16.1-Hannover
QGIS code revision
b381a90dca
Compiled against Qt
5.11.2
Running against Qt
5.11.2
Compiled against GDAL/OGR
3.1.4
Running against GDAL/OGR
3.1.4
Compiled against GEOS
3.8.1-CAPI-1.13.3
Running against GEOS
3.8.1-CAPI-1.13.3
Compiled against SQLite
3.29.0
Running against SQLite
3.29.0
PostgreSQL Client Version
11.5
SpatiaLite Version
4.3.0
QWT Version
6.1.3
QScintilla2 Version
2.10.8
Compiled against PROJ
6.3.2
Running against PROJ
Rel. 6.3.2, May 1st, 2020
OS Version
Windows 10 (10.0)
Active python plugins
db_manager; 
MetaSearch; 
processing

**Additional context**

Laptop is running Windows 10 20H2. Fully patched. It has a quite recent, clean, bare-metal install of Windows.

Per https://issues.qgis.org/issues/17874, it appears it was believed that this was resolved.

**Comment by reyemtm** [comment]
I believe this is upstream in QT - the workaround as you mentioned at least is to create a manifest file - see [this example](https://danantonielli.com/wp-content/uploads/2014/03/manifest.txt) from [this post](https://danantonielli.com/adobe-app-scaling-on-high-dpi-displays-fix/). I forget to copy my file and every time I upgrade I have to create it again. Definitely a pain and I'm not too sure on what all is absolutely necessary in the manifest file.

also this looks to be a duplicate of https://github.com/qgis/QGIS/issues/20008

**Comment by arencambre** [comment]
Thank you. That manifest workaround appears to be superseded per Dan's April 8, 2017 comment: https://danantonielli.com/adobe-app-scaling-on-high-dpi-displays-fix/#comment-16689.

**Comment by reyemtm** [comment]
Changing the high dpi settings did not work for me. Still using the manifest file.

**Comment by DantSu** [comment]
Here a batch code to create regedit key and auto generate manifest file.
Put this code to a .bat file and execute it as administrator.

```batch
@echo off
chcp 65001
setLocal EnableDelayedExpansion

reg add "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\SideBySide" /v PreferExternalManifest /t REG_DWORD /d 0x00000001 /f


set /p path="Please, enter full path to the .exe file:"

echo "%path%"

>"%path%.manifest" (
    echo:^<?xml version=^"1.0^" encoding=^"UTF-8^" standalone=^"yes^"?^>
    echo:
    echo:^<assembly xmlns=^"urn:schemas-microsoft-com:asm.v1^" manifestVersion=^"1.0^" xmlns:asmv3=^"urn:schemas-microsoft-com:asm.v3^"^>
    echo:
    echo:^<dependency^>
    echo:  ^<dependentAssembly^>
    echo:    ^<assemblyIdentity
    echo:      type=^"win32^"
    echo:      name=^"Microsoft.Windows.Common-Controls^"
    echo:      version=^"6.0.0.0^" processorArchitecture=^"*^"
    echo:      publicKeyToken=^"6595b64144ccf1df^"
    echo:      language=^"*^"^>
    echo:    ^</assemblyIdentity^>
    echo:  ^</dependentAssembly^>
    echo:^</dependency^>
    echo:
    echo:^<dependency^>
    echo:  ^<dependentAssembly^>
    echo:    ^<assemblyIdentity
    echo:      type=^"win32^"
    echo:      name=^"Microsoft.VC90.CRT^"
    echo:      version=^"9.0.21022.8^"
    echo:      processorArchitecture=^"amd64^"
    echo:      publicKeyToken=^"1fc8b3b9a1e18e3b^"^>
    echo:    ^</assemblyIdentity^>
    echo:  ^</dependentAssembly^>
    echo:^</dependency^>
    echo:
    echo:^<trustInfo xmlns=^"urn:schemas-microsoft-com:asm.v3^"^>
    echo:  ^<security^>
    echo:    ^<requestedPrivileges^>
    echo:      ^<requestedExecutionLevel
    echo:        level=^"asInvoker^"
    echo:        uiAccess=^"false^"/^>
    echo:    ^</requestedPrivileges^>
    echo:  ^</security^>
    echo:^</trustInfo^>
    echo:
    echo:^<asmv3:application^>
    echo:  ^<asmv3:windowsSettings xmlns=^"http://schemas.microsoft.com/SMI/2005/WindowsSettings^"^>
    echo:    ^<ms_windowsSettings:dpiAware xmlns:ms_windowsSettings=^"http://schemas.microsoft.com/SMI/2005/WindowsSettings^"^>false^</ms_windowsSettings:dpiAware^>
    echo:  ^</asmv3:windowsSettings^>
    echo:^</asmv3:application^>
    echo:
    echo:^</assembly^>
)

echo "Manifest file generated !"

pause >nul
```

**Comment by jbp35** [comment]
I am facing the same issue and I think that I have been able to track partially the origin. It appears to me that currently QGIS does not take into account the devicePixelRatio as it can be seen in this comparison of QGIS console output and the standalone DprGadget provided by QT:

![Screenshot 2022-08-20 090325](https://user-images.githubusercontent.com/19955142/185874273-9fb62e08-3ebd-48b3-9c8c-ca2d3d67345a.jpg)

The UI scaling issue can be fixed by adding the following lines of code to https://github.com/qgis/QGIS/blob/master/src/app/main.cpp after line 918 

```
// Set up for high DPI displays
QCoreApplication::setAttribute(Qt::AA_EnableHighDpiScaling, true);
QCoreApplication::setAttribute(Qt::AA_UseHighDpiPixmaps);
QGuiApplication::setHighDpiScaleFactorRoundingPolicy(Qt::HighDpiScaleFactorRoundingPolicy::PassThrough);
```

Note: @nyalldawson  this can be of interest as I think that QT6 set these attributes by default. 

Alternatively, it is also possible to run qgis from a batch file with following commands to test: 

```
set QT_ENABLE_HIGHDPI_SCALING=1 
set QT_SCALE_FACTOR_ROUNDING_POLICY=PassThrough
C:\OSGeo4W\bin\qgis-bin.exe
```

However, it is not possible to set UseHighDpiPixmaps from command line so the icons will look blurry. 

After doing this, UI looks great on my device. However, there seems to be an issue with the scale of the map on the high DPI screen which is 2X too big. I have not been able to figure out where this comes from, but it looks like QgsMapSettings currently use a default devicePixelRatio of 1 which is never been updated. 

Before:
![before](https://user-images.githubusercontent.com/19955142/185874338-3b0b33de-f255-414e-9bed-5ad2aea94e7a.jpg)


After:
![after](https://user-images.githubusercontent.com/19955142/185874311-433a1ea4-c1ff-4525-8940-72b30ceb375c.jpg)

**Comment by GvdDool** [comment]
Apologies for stepping in late, but I am facing similar problems on my Windows 13'' laptop with full HD1080. 

I don't know enough about fixing/editing the code base, so I wouldn't know where to change:
// Set up for high DPI displays
QCoreApplication::setAttribute(Qt::AA_EnableHighDpiScaling, true);
QCoreApplication::setAttribute(Qt::AA_UseHighDpiPixmaps);
QGuiApplication::setHighDpiScaleFactorRoundingPolicy(Qt::HighDpiScaleFactorRoundingPolicy::PassThrough); 

and test if this solved my problem of having a status bar not behaving as expected (related conversation on: https://gis.stackexchange.com/questions/457210/qgis-gui-layout-and-screen-scaling-in-windows?noredirect=1#comment746623_457210 

I hope someone can direct me to some resources to fix the GUI. Thanks.

**Comment by GvdDool** [comment]
Thanks jbp35,
From your earlier comment: 

> However, it is not possible to set UseHighDpiPixmaps from command line so the icons will look blurry. 

Is there a workaround for this, or should I try and see how blurry the icons are after setting the parameters?

**Comment by jbp35** [comment]
> Thanks jbp35, From your earlier comment:
> 
> > However, it is not possible to set UseHighDpiPixmaps from command line so the icons will look blurry.
> 
> Is there a workaround for this, or should I try and see how blurry the icons are after setting the parameters?

You can give it a try, anyway these settings will only apply when you launch QGIS using the .bat file.

**Comment by GvdDool** [comment]
Thanks, jbp35, but I think I will try to downgrade to my previous version first, and I hope that by doing so I can normally work again before using scripts to workaround a problem introduced by upgrading to the latest version.

**Comment by nyalldawson** [comment]
Should be all fixed in QGIS 3.32 -- please open new tickets for any specific individual hidpi issues remaining after this release
